framework/Optimizers/stepManipulators/ConjugateGradient.py

```diff
@@ -0,0 +1,436 @@
+      lastStepInfo['persistence'] = 0
+      lastStepInfo['prev task'] = 'CONV'
+      lastStepInfo = self._startLineSearch(lastStepInfo, curPoint, curObjVal, curGrad, curGradMag)
+      stepSize, task = self._lineSearchStep(lastStepInfo, curObjVal)
+    elif task.startswith((b'WARN', b'ERROR')):
+      if task.startswith(b'WARN'):
+        lastStepInfo['prev task'] = 'WARN'
+        msg = task[9:].decode().lower()
+        print('ConjugateGradient WARNING: "{}"'.format(msg))
+      elif task.startswith(b'ERROR'):
+        lastStepInfo['prev task'] = 'ERROR'
+        print('ConjugateGradient ERROR: Not able to continue line search!')
+      lastStepInfo['persistence'] += 1
+      if lastStepInfo['persistence'] >= self._persistence:
+        raise NoMoreStepsNeeded
+    else:
+      self.raiseAnError(RuntimeError, 'Unrecognized "task" return from scipy.optimize.minpack2: "{}"'.format(task))
+
+    lastStepInfo['stepSize'] = stepSize
+    lastStepInfo['task'] = task
+
+    currentPivot = lastStepInfo['pivot']['point']
+    newPivot = currentPivot - stepSize * lastStepInfo['pivot']['gradient']
+    newOpt = dict((var, newPivot[v]) for v, var in enumerate(self._optVars))
+    return newOpt, stepSize, lastStepInfo
+
+  def fixConstraintViolations(self, proposed, previous, fixInfo):
+    """
+      Given constraint violations, update the desired optimal point to consider.
+      @ In, proposed, dict, proposed new optimal point
+      @ In, previous, dict, previous optimal point
+      @ In, fixInfo, dict, contains record of progress in fixing search
+      @ Out, proposed, new proposed point
+      @ Out, stepSize, new step size taken # TODO need?
+      @ Out, fixInfo, updated fixing info
+    """
+    # TODO this is copied from GradientHistory; it should be updated for the ConjugateGradient when
+    #      we know how we want to do this
+    # DESIGN
+    # While not okay:
+    # 1. See if cutting the step will fix it.
+    # 2. If not, try rotating towards a random perpendicular. Repeat 1.
+    # 3. If not, try a new random perpendicular. Repeat 1. Repeat N times.
+    # TODO should this be specific to step manipulators, or something else?
+    # TODO updating opt point in place! Is this safe?
+    minStepSize = fixInfo['minStepSize']
+    stepVector = dict((var, proposed[var] - previous[var]) for var in self._optVars)
+    stepDistance, stepDirection, _ = mathUtils.calculateMagnitudeAndVersor(list(stepVector.values()))
+    if 'originalStepSize' not in fixInfo:
+      fixInfo['originalStepSize'] = stepDistance
+    if 'perpDir' in fixInfo:
+      perpDir = fixInfo['perpDir']
+    # if not done cutting step, start cutting
+    if stepDistance > minStepSize:
+      # cut step again
+      stepSize = 0.5 * stepDistance # TODO user option?
+      for v, var in enumerate(stepVector):
+        proposed[var] = previous[var] + stepSize * stepDirection[v]
+      print(' ... cutting step ...') # norm step to {}, new norm opt {}'.format(stepSize, proposed))
```

Comment:
printing ?

Reply:
no access to a message handler. This could be changed if we change our policies about APIs having message handlers (e.g. CodeInterface classes).